[WPE][GTK] createNewPage in WebKitUIClient.cpp needs some hooking up of the API::PageConfiguration

mcatanzaro opened the pull request:
#### 4f5aa1a1182dc07afb17a676cf68ee07e5843179
<pre>
[WPE][GTK] createNewPage in WebKitUIClient.cpp needs some hooking up of the API::PageConfiguration
<a href="https://bugs.webkit.org/show_bug.cgi?id=273975">https://bugs.webkit.org/show_bug.cgi?id=273975</a>

Reviewed by Darin Adler and Alex Christensen.

This resolves a FIXME where the API::PageConfiguration gets ignored when
creating a new related web view. The control flow isn&apos;t obvious unless
you&apos;re familiar with GTK/WPE ports, but the WebKitWebView emits the
WebKitWebView::create signal, then the application using WebKit is
supposed to create a new web view using
webkit_web_view_new_with_related_view().

Note that some members of the PageConfiguration are always taken from
the related view rather than the PageConfiguration we receive.

This commit additionally adjusts the sanity check in
webkitWebViewCreateNewPage to match the corresponding Cocoa checks in
UIDelegate::UIClient::createNewPage, so it (hopefully) won&apos;t fail when
site isolation is enabled. This was suggested by Carlos Garcia.

* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewConstructed):
(webkitWebViewCreateNewPage):
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:

Canonical link: <a href="https://commits.webkit.org/278896@main">https://commits.webkit.org/278896@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d6d4fc204b7da282ae5f8c09cc56ef81b1fcdf4b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51851 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31163 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4215 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55119 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2543 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37538 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2252 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/42175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/53950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28777 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44723 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23311 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/26059 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/48014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2077 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56710 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26972 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/1959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49577 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/28209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44824 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11333 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29106 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/27946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
```diff
@@ -2023,7 +2023,7 @@ WebProcessPool& webkitWebContextGetProcessPool(WebKitWebContext* context)
+    auto pageConfiguration = relatedView ? webkitWebViewGetConfigurationForNextRelatedView(relatedView) : API::PageConfiguration::create();
     pageConfiguration->setProcessPool(context->priv->processPool.get());
     pageConfiguration->setPreferences(webkitSettingsGetPreferences(webkit_web_view_get_settings(webView)));
     pageConfiguration->setRelatedPage(relatedView ? &webkitWebViewGetPage(relatedView) : nullptr);
```
Can webkitWebContextCreatePageForWebView be called with a null relatedView?

I think these page configuration setters are duplicate work if relatedView is non-null because the values area already set in WebPageProxy::createNewPage, but I don't think they hurt.  I also can't say I'm completely following all possible code flows here.

carlosgcampos replied:
Reviewing this I've realized that page configuration creation doesn't really belong to web context, so I've created https://github.com/WebKit/WebKit/pull/28383 to move it to the web view which will simplify this patch too.

mcatanzaro replied:
> Can webkitWebContextCreatePageForWebView be called with a null relatedView?

Yes, and that case was broken and would trigger the assert. Good catch.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
```diff
@@ -2023,7 +2023,7 @@ WebProcessPool& webkitWebContextGetProcessPool(WebKitWebContext* context)
 void webkitWebContextCreatePageForWebView(WebKitWebContext* context, WebKitWebView* webView, WebKitUserContentManager* userContentManager, WebKitWebView* relatedView, WebKitWebsitePolicies* defaultWebsitePolicies)
 {
-    auto pageConfiguration = API::PageConfiguration::create();
+    auto pageConfiguration = relatedView ? webkitWebViewGetConfigurationForNextRelatedView(relatedView) : API::PageConfiguration::create();
```
Note that user can create a web view with a related view without being in create callback, so we can have a related view but not a page configuration.

mcatanzaro replied:
Yes, this is what Alex was concerned about. Thanks.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -287,6 +287,7 @@ struct _WebKitWebViewPrivate {
     HashSet<unsigned> frameDisplayedCallbacksToRemove;
 #endif
 
+    RefPtr<API::PageConfiguration> configurationForNextRelatedView;
```
This is weird, but I don't think there's a better way to do it. Maybe we can find a better name.

mcatanzaro replied:
It's the best name I can think of.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -2598,6 +2602,12 @@ RefPtr<WebPageProxy> webkitWebViewCreateNewPage(WebKitWebView* webView, const Wi
+    ASSERT(webView->priv->configurationForNextRelatedView);
+    return *webView->priv->configurationForNextRelatedView;
+}
+
```
We don't need this after https://github.com/WebKit/WebKit/pull/28383 lands. With that patch we can do something like this in `webkitWebViewConstructed`:

```
Ref<API::PageConfiguration> pageConfiguration = priv->relatedView && priv->relatedView->priv->configurationForNextRelatedView ? *priv->relatedView->priv->configurationForNextRelatedView : webkitWebViewCreatePageConfiguration(webView);
webkitWebViewCreatePage(webView, WTFMove(pageConfiguration));
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -2581,8 +2582,11 @@ void webkitWebViewSetIcon(WebKitWebView* webView, const LinkIcon& icon, API::Dat
 {
+    ASSERT(!webView->priv->configurationForNextRelatedView);
+    SetForScope configurationForNextRelatedViewGuard(webView->priv->configurationForNextRelatedView, WTFMove(configuration));
+
```
The received configuration is a copy of the related page configuration only changing the opener process when site isolation is enabled. So, in this function we can add another check to make sure the opener process matches too.

mcatanzaro replied:
I don't understand. The opener process is the only thing that's different, so we should check that it's not different?

mcatanzaro replied:
I misunderstood this. The opener process is the web process that calls window.open(), triggering creation of the new web process. The opener process of the new web view should match that specified by the API::PageConfiguration, or we know the application using WebKit didn't create the web view properly.

darinadler commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -2646,21 +2648,27 @@ void webkitWebViewSetIcon(WebKitWebView* webView, const LinkIcon& icon, API::Dat
     webkitWindowPropertiesUpdateFromWebWindowFeatures(newWebView->priv->windowProperties.get(), windowFeatures);
 
-    return &getPage(newWebView);
+    return newPage;
```
Since this is converting a `Ref` to a `RefPtr`, I think we might save one reference count churn by writing `WTFMove`, even though that would be wrong if the return type was `Ref`.

mcatanzaro replied:
I'll try it.

mcatanzaro replied:
Works.

mcatanzaro replied:
Well, it worked for me only because I tested with Clang, which doesn't have -Wredundant-move or is more lenient about it. GCC doesn't like it. I'll revert this before landing.

mcatanzaro replied:
Um, that must mean that it uses the rvalue constructor automatically (otherwise it wouldn't be "redundant"). It must be the guaranteed copy elision, so I think there should be no refcount churn.

darinadler replied:
Some day I will learn these rules.

darinadler commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -909,7 +910,8 @@ static void webkitWebViewConstructed(GObject* object)
         priv->websitePolicies = adoptGRef(webkit_website_policies_new());
 
-    webkitWebViewCreatePage(webView, webkitWebViewCreatePageConfiguration(webView));
+    Ref<API::PageConfiguration> configuration = priv->relatedView && priv->relatedView->priv->configurationForNextRelatedView ? priv->relatedView->priv->configurationForNextRelatedView.releaseNonNull() : webkitWebViewCreatePageConfiguration(webView);
```
Can we just write `Ref` without the template argument?

mcatanzaro replied:
Probably...

mcatanzaro replied:
Works.

mcatanzaro committed:
[WPE][GTK] createNewPage in WebKitUIClient.cpp needs some hooking up of the API::PageConfiguration
https://bugs.webkit.org/show_bug.cgi?id=273975

Reviewed by Darin Adler and Alex Christensen.

This resolves a FIXME where the API::PageConfiguration gets ignored when
creating a new related web view. The control flow isn't obvious unless
you're familiar with GTK/WPE ports, but the WebKitWebView emits the
WebKitWebView::create signal, then the application using WebKit is
supposed to create a new web view using
webkit_web_view_new_with_related_view().

Note that some members of the PageConfiguration are always taken from
the related view rather than the PageConfiguration we receive.

This commit additionally adjusts the sanity check in
webkitWebViewCreateNewPage to match the corresponding Cocoa checks in
UIDelegate::UIClient::createNewPage, so it (hopefully) won't fail when
site isolation is enabled. This was suggested by Carlos Garcia.

* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewConstructed):
(webkitWebViewCreateNewPage):
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:

Canonical link: https://commits.webkit.org/278896@main
